c/_executormodule.c

```diff
@@ -359,24 +351,64 @@ static int translate_python_exception(lua_State *L, PyGILState_STATE gstate) {
 
 void store_python_capsule(lua_State *L,
                           PyObject* val,
-                          long cycle_key,
                           int should_cache,
                           int recursive,
                           int raw_lua_args) {
-    // our caller already added us to cycles so we don't have to worry about
-    // it here
     lua_capsule* capsule =
         (lua_capsule*)lua_newuserdata(L, sizeof(lua_capsule));
 
-    // we don't hold the GIL so we can't do anything but store pointers. note
-    // that we explicitly are not refcounting here: we rely on `cycles` to keep
-    // us live
     capsule->val = val;
-    capsule->cycle_key = cycle_key;
     capsule->cache_ref = LUA_REFNIL; // cache is populated lazily
     capsule->cache = should_cache;
     capsule->recursive = recursive;
     capsule->raw_lua_args = raw_lua_args;
+
+    // assign the metatable of the userdata to get the methods
+    lua_getfield(L, LUA_REGISTRYINDEX, EXECUTOR_LUA_CAPSULE_KEY);
+    lua_setmetatable(L, -2);
+
+    // we don't do explicit refcounting. instead we rely on our references dict
```

Comment:
The old comment mentioned that we don't hold the GIL for this block. Is that still true?

Reply:
No, it's not true anymore. We now retain the GIL from both call sites